h2o-core/src/main/java/water/fvec/Frame.java

```diff
@@ -71,14 +71,14 @@
    * Given a temp Frame and a base Frame from which it was created, delete the
    * Vecs that aren't found in the base Frame and then delete the temp Frame.
    *
-   * TODO: Really should use Scope but Scope does not.
    */
   public static void deleteTempFrameAndItsNonSharedVecs(Frame tempFrame, Frame baseFrame) {
-    Key[] keys = tempFrame.keys();
-    for( int i=0; i<keys.length; i++ )
-      if( baseFrame.find(keys[i]) == -1 ) //only delete vecs that aren't shared
-        keys[i].remove();
-    DKV.remove(tempFrame._key); //delete the frame header
+    Set<Key> baseFramekeys = new HashSet<>(baseFrame._keys.length);
+    for (int i = 0; i < baseFrame._keys.length; i++) {
+      baseFramekeys.add(baseFrame._keys[i]);
+    }
+    
+    tempFrame.retain(new Futures(), baseFramekeys);
```

Comment:
This usually leads to very strange behavior, race condition and failing tests. You do need to wait for completion of Futures _unless_ it is a responsibility of the caller. In this case it is not.

I suggest to leave this method as-is, and use the new method when it proves itself.